### DISCUSSION TITLE
Release process

### DISCUSSION THREAD

**Discussion by carlos-granados** [Behat/Behat#1484] [General]
@everzet @acoulton I have opened this discussion to talk about our release process. @everzet your experience here will be very valuable.
- When should we create a new release? Looking at the history of Behat, releases seemed to happen not very often, maybe every 6 months of so. In my experience with other tools like Rector, users are happier if releases happen more frequently, every 2-3 weeks but perhaps they see more action. Maybe minor releases could happen less often, same cadence as before but do patch releases more often?
- Who will take care of creating the releases? Should there be a person in charge?
- How is the changelog handled? Is there a tool to build it? 
- I can see that phar building happens as part of the tests and it is published when a release is done. Is there anything else needed to complete the release?
- Is the release announced anywhere (twitter, etc...). If so, how is this handled?

**Comment by acoulton** [comment]
Thanks @carlos-granados I was going to message you something similar!

My thoughts:

* We should release as often as possible (when there's something to release). No more than 1-2 weeks after merging a PR, sooner (maybe even straight away) if the PR contains a bugfix. It's frustrating as a contributor/user of a project to see a contribution merged but then have to switch to a fork / dev branch to use it for an arbitrary period before it gets released. And the smaller & more often we release, the more chance of people finding any issues while the people who contributed changes are still around to investigate & fix them.
* We should jointly take care of releases, with a similar policy to PRs that 2 maintainers (e.g. you & me) approve a release.
* It looks to me like the changelog is currently a manual process - there's an issue #1400 to automate it.

I don't know the answers to your last two (is there anything else needed to release? is it announced anywhere?) but I think the answer is "no"?

I have an idea for a workflow that could deal with changelog / approval / release in a way that would be efficient and reliable enough for us to run it regularly, but I've got to run now. I'll come back to you on that later.

**Comment by stof** [comment]
The release process itself looks like that:

1. update the changelog (manually), even though it looks like the latest release has stopped updated the file in the repo, relying only on the github release to hold it (but still needs to be computed anyway)
2. create a git tag. those tags use the `v` prefix (which is optional as far as composer is concerned) to play nicer with completion in git commands as it keeps them separate from branches (so `git switch 4<Tab>` can complete the `4.x` branch name directly)
3. push the tag
4. create the GitHub release associated with the tag, with its changelog

Publishing the phar for the release is automated.

**Comment by acoulton** [comment]
Thanks @stof 

My proposal for a future workflow (which I think I can make work, but obviously we'd need to test it) is:

#### On every merge to master

A github action:

* Generates a changelog entry and inserts it in the correct place in the CHANGELOG file (which I would like to see grouped by features / fixes / internal (e.g. chores, tests, refactoring, etc) - we can use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) prefixes when merging PRs to make this easier to automate).
* Commits that CHANGELOG to a `next-3.x` branch and merges master into it (or rebases it onto master - will need to see what works best for us)
* Creates a "Next release" pull request to merge `next-3.x` back to master (if one doesn't already exist)
* Sets (or updates) the PR title to include the proposed version number - bumping the minor / patch version depending on the commits included.

#### Preparing for a release

Maintainers (and anyone else) will at any time be able to use the "Next release" pull request and the CHANGELOG diff it contains to:

* See what / how much is ready to be released to see whether it's worth releasing yet
* Check if the automatic CHANGELOG properly describes the upcoming changes, and discuss / propose / make changes as required
* Link any other issues or PRs that we decide are blockers
* Use the existing PR review tools to co-ordinate whether / when we agree the release is ready to go

#### Releasing

Once 2 maintainers have approved the "next release" PR, any maintainer can merge it to master. We'd probably do this as a squash merge to keep the history clean, since the actual merge would only be modifying the CHANGELOG.

A github action will detect this merge and:
* create the github tag with the appropriate version number
* create the github release, and attach the release notes from the CHANGELOG
* deletes the `next-3.x` branch and creates a new one from current master.

Github actions will then publish the phar and attach it to the release using the existing workflow.

#### Major version bumps

If/when we're ready to start work on a new major version, the same flow above would work - we'd manually decide to merge a PR to e.g. the `4.x` branch. The workflows would detect that and run the process using a `next-4.x` branch (and "Next 4.x release" PR) so we could be working towards 3.x and 4.x releases in parallel and release them at different times.

### Advantages

Everything in this flow is either automated (and repeatable), or can be done from the github web or mobile interfaces, and each step in the process only requires one physical action by a maintainer to trigger it. It also gives us a clear & trackable process to co-ordinate releases without too much effort. And it should make it easy to publish a changelog that is actually meaningful & useful to end-users.

@carlos-granados what do you think? If you agree this is worth trying, I can build a first version of the github actions for us to experiment with and see how it works in practice.

**Reply by uuf6429** [reply]
Some thoughts on this:

I'm of the opinion of just having a `3.x` branch - I'd even remove the `master` branch and make the 3.x one as the dev/default branch (and change it again every time there's a major version bump).

We should be squashing PRs into one commit, but not an entire release (into one commit) - that would make history essentially useless (if I understood well).

I'm not a fan of having a changelog file - I've never had to or bothered reading one in my entire career, and I feel it's the same for many people. GitHub and git version history takes care of that. It would also mean less things to bother automating for a release.

My normal workflow for a release is relatively simple - I just draft a new release from the GitHub web ui. This takes care of pretty much everything - creating a tag, description for the release (which can be reviewed and edited) etc.

**Reply by acoulton** [reply]
@uuf6429 

> just having a 3.x branch instead of master

I'd be in favour of renaming `master` to `3.x`, as a separate thing. However what I'm proposing is a separate branch as a way to have a PR that shows what is *already on master/3.x* but is waiting to be released.

> We should be squashing PRs into one commit, but not an entire release (into one commit)

Some PRs can be squashed, others can't - it depends on the content. But I'm not proposing squashing a release - the squash merge I mentioned above would only be for the CHANGELOG - everything else would be in history as usual.

> I'm not a fan of having a changelog file - I've never had to or bothered reading one in my entire career, and I feel it's the same for many people. GitHub and git version history takes care of that. It would also mean less things to bother automating for a release.

I don't want to waste your (or my) time with a [someone is wrong on the internet](https://xkcd.com/386/) 🤣 but I completely disagree. I spend a lot of working time reading change histories, and some projects make that incredibly hard.

Git version history is for developers, a CHANGELOG is for users. When the users are also developers, it's easy to forget that - but the information that **developers using a product** need is not the same as **developers working on a product** need. Keeping a CHANGELOG (and proper release notes) forces people to think about what they need to tell end-users, rather than assuming everyone will understand internals enough to know how a release might affect them.

**Reply by uuf6429** [reply]
> I'd be in favour of renaming master to 3.x, as a separate thing. However what I'm proposing is a separate branch as a way to have a PR that shows what is already on master/3.x but is waiting to be released.

So your objective is to have a PR that multiple people can review, correct? Sadly that's not something GitHub's release process offers (AFAIK).

> I mentioned above would only be for the CHANGELOG - everything else would be in history as usual.

Good, but then it's a bit unclear to me how that would work... do you mean there would be one commit for the changelog, or that the changelog changes are squashed into whichever last commit happened to be in the release?

> I spend a lot of working time reading change histories, and some projects make that incredibly hard.

- Is that really relevant though? I'm reading change history as well ... but I've never _really_ opened `vendor/<name>/<project>/CHANGELOG.md`. I can confidently say many projects have such a file "just because" and they don't really maintain it. Which means as a user, I know better not to trust that file and go directly to the source.. Do we really want to go into the extra effort and responsibility of maintaining such a file?
- I obviously meant this from a developer point of view. I'm not saying changelogs are universally useless, just that in terms of GitHub releases, they are - they tend to contain less useful information while requiring a lot of additional manual work.
- Issues should always be readable and understandable. The users searching in a change history are also the ones searching and/or reporting bugs... so I say fix the problem upstream - make sure issues are well titled.
- I get the feeling that having a changelog file in the PHP community is from cargo cult mentality... symfony has one, so we must have one too (that goes for a lot of other projects).
- Speaking of which, compare [this](https://github.com/symfony/symfony/blob/7.2/CHANGELOG-7.0.md) with [this](https://github.com/symfony/symfony/releases).

Anyway, that's from my point of view as a package user, I know not to trust a _changelog file_ - at best they lie and at worst they're half-baked - either case they're neither reliable nor valuable.

Obviously you don't have to convince me otherwise, we can keep the changelog if that's what we want.

**Reply by acoulton** [reply]
@uuf6429 I think there's some misunderstandings about what I'm proposing that would be easier to see if I just get a prototype up and running to demonstrate.

I would say I've been reading & writing changelogs since long before symfony so it's definitely not a cargo cult for me, and also that I'm not sure what you were trying to illustrate as the comparison between the two symfony links, but I don't think either of them is a good example.

**Comment by carlos-granados** [comment]
@acoulton thanks for your proposal, I'll take a more detailed look this weekend. In the meantime, what about using something like what Rector uses to create their release notes:
https://github.com/rectorphp/release-notes-generator
It creates some really nice release notes:
https://github.com/rectorphp/rector/releases

**Comment by acoulton** [comment]
@carlos-granados yes I've seen theirs & was going to take a look at their tooling - there are some other projects / existing github actions that are interesting too. However:

* All the automated tools rely heavily on commits and pull-requests being well titled. It's very common to see changelogs that might make sense to people involved in the project but are hard to understand for end-users. Being disciplined about renaming PRs that are unclear before merging, and/or taking care with the wording of merge commits rather than just using the github default can help, but isn't guaranteed. Even in that rector example you've linked, there are things described as "bugfixes" that are actually just code style and variable naming internal to classes that will make no difference to execution of the tool and really should be grouped separately.
* Sometimes the PR title alone is enough, sometimes even if it's well titled the changes really need another line to summarise how they affect people. Otherwise you end up with release notes that leave users having to click into multiple PRs and read the full description (or worse, the diff) to know if they're relevant. That's not a decision that can be automated, so there needs to be an opportunity for humans to edit - and ideally review - the changelog between generation and release. That's much easier & more effective if it can be done as soon as the work is merged, rather than being left to one person at the very end of the release process.
* Similarly I think there needs to be a traditional CHANGELOG file as well as the github releases. It's incredibly annoying if you're bumping more than one version at once and the only way to find the information you need is to find and read each release in turn on github rather than just skim-reading your way up a single file. Especially if there are lots of small releases, which is very good for most purposes, but very hard if the release notes are all separate.

People bumping dependencies are often bumping more than one thing at once, and sometimes more than one version at a time, and across multiple projects (especially if e.g. they are switching to a new major series / PHP language version / other dependency that has been holding things back). I think it's important to make that as easy as possible, by making the changes really clear & quick to understand.

My experience in work over the last few years has been that the quality of change information has declined dramatically (across all languages / platforms). Which I think is because projects have automated their release processes and left release notes as an afterthought, then tried to automate a solution for that. It's increasingly common to find changelogs that are literally no different to reading the git oneline commit history, and therefore serve no real purpose. Personally, I think that's also partly why we're seeing more and more supply chain attacks creeping in, because people have become overwhelmed trying to see what's changed and have just given up and done whatever composer/dependabot/whatever told them.

I'd really like us to try and find a process that doesn't create a huge burden for us, but helps us make life easier for end-users :)

**Comment by stof** [comment]
I've never automated release notes in my own projects precisely because it forces to actually think about those release notes at the time of releasing.

**Reply by acoulton** [reply]
@stof same - that's why I'd like to experiment with a half-way house: automate generating a draft as things are merged, but still have a process that forces us to think about them (and edit if necessary) as part of release.

**Comment by carlos-granados** [comment]
@acoulton I reviewed your proposed workflow in more detail and it seems very reasonable, if you could start working on a possible implementation that would be great. Please let me know if I can help you in any way

**Comment by ciaranmcnulty** [comment]
Re: the cadence of releases it's always been 'as often as possible, as maintainer energy allowes'. We have just locked that energy :)

As long as we are sticking to SemVer I'd be happy with a release per merge

@ocramius has some highly-automated release processes we could look at / steal (that AFAIK rely on Milestones in GH)

**Reply by Ocramius** [reply]
Ya, I now release minor versions at every improvement (even just docs), and only major/breaking changes require a bit more caution.

**Reply by acoulton** [reply]
Thanks - I'll take a good look at those. Some milestone-based workflows I've seen feel like they'd work better with a single project expert / benevolent dictator calling the shots but might be a bit harder for the more collaborative model @carlos-granados and I are aiming for.

We're definitely sticking to SemVer, and I'd ideally like to see more structured differentiation of patch vs point numbering (rather than everything just being 3.x.0 by default).

I agree with "as often as possible" but I don't think we want to be tied to "release per merge". If things are quiet and there's only one PR in the pipeline then I agree, no point waiting. But if there's a few different things likely to be merged in the same few days, my experience as a user is it's often more efficient to see that as a single release.

As a user, I ideally wouldn't want to see a version number bump (even a patch) if it only contains a docs improvement - that just becomes noise and wasted effort in the dependency review/update process at my end.

But it's a good point that if we do want to immediately release a PR (which has already been reviewed and approved by multiple maintainers), we don't want a workflow that adds too many steps to that if the PR itself has already been reviewed and approved. I'll give that some thought.

**Reply by stof** [reply]
to me, the documentation needs to be deployed on each push to the branch, not only on releases, which solves the case of not needing to do documentation-only releases.

**Reply by Ocramius** [reply]
FWIW, I generally close non-major milestones whenever they have no pending issues: small incremental upgrades really help users with moving forward, even if noisy.

For the collaboration side, any blocker for a release can be an issue assigned to the milestone and not closed: it will prevent an accidental release. Laminas has been auto-releasing for a while, and it's a project with lots of contributors.

For the release notes: the entire milestone text is used as release prefix markdown.

**Reply by acoulton** [reply]
@stof absolutely - my point was less about getting the docs published, more that we don't want a workflow that *forces* a release every time a PR is merged, as there might be changes that are purely docs (or internal stuff e.g. github actions bumps) that don't need to bother users straight away. TBH at the moment docs is a moot point since they are not in this repo / release flow anyway.

**Reply by acoulton** [reply]
@Ocramius I'm planning to have a good look at Laminas, my big question is do you decide (alone) when / whether to close a milestone? That's the bit where Carlos and I were envisaging collaboration, hence thinking a PR was an easier way to co-ordinate since there are comments and approvals already built-in to that flow. But maybe if we are collaborating on PR merges it's OK for one of us to decide when we feel like releasing.

**Reply by Ocramius** [reply]
> do you decide (alone) when / whether to close a milestone?

Any of the maintainers can do any release at any time: there needs to be some trust in each other, otherwise you wouldn't be co-maintainers. Major releases are discussed a lot, minor releases are no-brainers.

>  That's the bit where Carlos and I were envisaging collaboration, hence thinking a PR was an easier way to co-ordinate since there are comments and approvals already built-in to that flow.

The tool I built shifts release focus from individual PRs to milestones: if you need things to still be done in a release, make an issue or a PR under that milestone, and that's it :) 

>  But maybe if we are collaborating on PR merges it's OK for one of us to decide when we feel like releasing.

My suggestion here is to always keep releasing, even if you just changed a single conditional, or made a tiny optimization: while your testing is in place, you also have a lot of downstream users that can help you pinpoint regressions better, if something goes awry (and it always does in software, even with 100% mutation test coverage and 100% type coverage).

**Reply by acoulton** [reply]
Thanks @Ocramius really value your thoughts.

**Comment by carlos-granados** [comment]
@acoulton while you work on the new release workflow, do you think it might be worth to do a one time "manual" release. We have merged a lot of stuff and it would be good to see it available for everyone. What do you think? We can then work on the new workflow for any new stuff that we add from this point onwards

**Reply by acoulton** [reply]
@carlos-granados I had the same thought and actually started work on a manual changelog for a release this morning, will finish it later today for you to review.

**Comment by carlos-granados** [comment]
@acoulton turns out that the `behat` version is hardcoded in the code: https://github.com/Behat/Behat/blob/0e2e07191d64b3146e252a2b69dd22fd052abe13/src/Behat/Behat/ApplicationFactory.php#L50
So right now when you run `behat -V` it reports `behat 3.13.0` 🤦 
I think that the best thing is that this is changed automatically in this file as part of the release process, what do you think?

**Reply by Ocramius** [reply]
Use Composer's built-in release detection for that: this sort of annoying release complication is what prompted me to write `ocramius/package-versions` in first place.

**Reply by acoulton** [reply]
@carlos-granados Urgh, that's annoying. 

Generally I'd agree that we should just use `ocramius/package-versions` for this (we'd need to make sure that composer in the phar build can identify the root package version).

The complication is that although we could safely rewrite `ApplicationFactory->getVersion()` to load a value at runtime, we can't do that for the `public const VERSION`. And removing that would be a BC break - if it was only likely to affect end-users I'd be tempted to just do it and highlight it in the CHANGELOG, but if it's used at all it is probably being used by extensions which would make it much harder for people to work around us removing it.

So I think there's two options:

* Keep updating that file for every release :scream: until the next major
* Keep the constant, but deprecate it and never update it again, and rewrite `ApplicationFactory->getVersion()` to get the *actual* value at runtime.

I'd suggest the second option is a good compromise?

**Reply by Ocramius** [reply]
The deprecation flow is what we went for in doctrine as well 👍
